### ISSUE TITLE
100% CPU & 530G VIRT memory used when opening excel file (specific file linked with demo)

### ISSUE THREAD

**Issue by ocelotsloth** [tealeg/xlsx#312]
## My Understanding of the Problem

I'm experiencing issues with opening a large excel file I have, which has been generated by https://microstrategy.gmu.edu (I'm fairly certain Microstrategy uses sheet.js). While the file size itself is only 2.6 MB, it does have around a half a million cells populated with data (this sheet contains every section of every course being taught at my university along with information about each).

The exact spreadsheet I am trying to open can be found at https://github.com/ocelotsloth/schedules/blob/master/dist/exampleFall2017.xlsx. 

When I try to open this file, the program hangs, consumes 100% CPU, and allocates itself a truly absurd amount of memory (or swap? To be honest I'm still not 100% clear on what VIRT is):

![hr9wn6d](https://user-images.githubusercontent.com/9255772/27938105-dbe6af86-6289-11e7-8b57-7d5c071a7fd2.png)

## Research I have done so far

I have done some research already and do not think it is related to either #109 or #114. This file is significantly smaller in filesize, and as explained below, the offending commit doesn't appear to be directly related to opening files (though I could very well be wrong on that second assertion).

This is a cross-platform issue that I have tested by compiling against Windows, Darwin, and Linux.

I performed a `git bisect` and was able to narrow down the offending commit to e07ef57987388ad2510dac5fa2b5dde5945b1b31, which was authored by @tealeg in October of 2015 (sorry to bring up such a distant memory). I've added the commit message below for quick reference:

    commit e07ef57987388ad2510dac5fa2b5dde5945b1b31
    Author: Geoffrey J. Teale <geoffrey.teale@canonical.com>
    Date:   Fri Oct 2 23:19:48 2015 +0200
    
        Populate the VMerge and HMerge properties of the cell based on the xlsxWorksheet.MergeCells.

This commit adds in some actions dealing with Merge Cells, but since I was more concerned with graduating high school than learning Golang at the time, I couldn't tell you the full extent of what it does and also don't know how fits in with other work being done at the time.

## A helpful repository to explore and test

Since the project I'm using this library in is actually my way of picking up and learning Golang, I'm well aware that I'm not really going to be of much further use to try and take a stab at the exact problem here. In lieu of that kind of help, I've gone ahead and made a simple program to clearly demonstrate the issue. 

`go get github.com/ocelotsloth/xlsx_issue_demo` should pull down both the demo program as well as the excel spreadsheet I am working with.

**Comment by tealeg** [comment]
@ocelotsloth I can't comment on the exact problem right now (I'm away on my summer holidays), but I can tell you that this is an extremely good and well thought out issue report.  If you keep up this level of work you're going to win a lot of friends!  I'll try to make some time to have a look at this for you.

**Comment by HeinOldewage** [comment]
Hi @ocelotsloth
I took a quick look, and it seems to be an interesting problem. 
I am testing on a windows machine, with the go master and go 1.6.

On my machine it is only using 180MB of ram (more than expected for a 2.6MB file). This is similar to your memory usage (VIRT is virtual memory, and is not really allocated. It is more like reserved, and is not really important for this problem).

The more interesting part for me is where the CPU time is going. Using a debugger (gdb) it narrowed it down to `lib.go:788` 
For context
```
decoder = xml.NewDecoder(rc)
error = decoder.Decode(sst)
```
That is the default `"encoding/xml"` package. So there is either something wrong with the xml of the sheet, or something wrong with go's xml package (there are 60 open issues on the main repo, but none of them seem related(but I only skimmed the surface)).

Have you encountered it with smaller worksheets?

**Comment by ocelotsloth** [comment]
@tealeg: Thank you for the kind words, this can definitely wait until you're not away.

@HeinOldewage thank you for taking a look! It appears you have a bit more experience with debugging than I, I'm happy you took the time.

>Have you encountered it with smaller worksheets?

I went ahead and tried the same testing program out with a far limited dataset (I just deleted almost all of the rows) and it worked perfectly. This leads me to believe that the excel sheet that is generated is perfectly valid.

I'm going to do some further testing to get a better profile, but I think that the issue at hand is more a performance issue having to do with this sheet than a specific error that is causing the program to cease functioning.

I find it curious that the breaking commit is in the area of merge cells, as this particular excel sheet has _*many*_ merged cells in both directions. I'd be the first to suggest that the data source I am working with verges on absurdity, but given the size of the dataset, it's the best I can do with what I have the ability to pull from microstrategy.gmu.edu.

I'm going to take a few minutes and generate a performance profile to see where and how quickly the library locks up.

**Comment by ocelotsloth** [comment]
So I left it on overnight with just the original spreadsheet, and it was actually able to load it entirely. It just took it 42 minutes to do so.

Is there something with that commit change that would adversely impact performance?

**Comment by tealeg** [comment]
@ocelotsloth well, there's obviously something there!  The first thing to say is that yes, go's encoding/xml package is inherently going to be quite heavy weight, we are after all loading the entire document into structs and exploding out some of the implications of the document.

The reason you can track it back to the specific commit you did is because your sheet has a (relatively) huge number of merged cell definitions, and that is the commit that added support for that functionality.  Prior to that commit your sheet would load quickly, because we simply wouldn't be doing anything with that data.  Unfortunately, it would also not read correctly in that version (AFAICT).

@HeinOldewage thanks for taking the time to help.

**Comment by tealeg** [comment]
If I had to guess I'd say that xlsxMergeCells.getExtent is the culprit.  For starters I can see an allocation in the loop that could end up sucking a lot of RAM for the number of merge cells in this sheet.

**Comment by ocelotsloth** [comment]
@tealeg thank you for pointing me in the right direction. I'm playing with the code and I think I have a better understanding of the issue.

Just to get a firm idea of the number of merge cell's in that sheet I added a print statement in that function to just print the `len` of `mc.Cells` and was suprised to see it was `105632105632` (approx 100 *billion*) merge cell definitions. Certainly not insignificant.

With a reduced set of the same excel sheet took the number of merge cell definitions down to `162`, and that function still ran `880` times. If it has to loop through all 162 cells defined in the merge cell struct, it's definitely going to cause such a slowdown.

Assuming a similar ratio with the larger data sheet, the number of iterations through that loop must be an incomprehensibly large number. @HeinOldewage I'd say that's where the CPU cycles are going.

I'm going to be taking my own look as I have time, but I think the solution would be to find a way to *(a)* reduce the number of times this function is required to run and/or *(b)* find a better way to address the merge cell we want besides looping through the entire array until it is found.

I can't imagine it would help the memory footprint, but I think a good solution would be to map the merge cells ahead of time in something that can be addressed more directly. Admittedly, now that I've actually looked at the way my data source constructed this excel sheet, but this still looks like an area that would benefit from some optimization. Even with 160 merge cells it still takes in the order of seconds to load the sheet.
